x-pack/libbeat/management/manager.go

```diff
@@ -0,0 +1,197 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/reload"
+
+	"github.com/satori/go.uuid"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/feature"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/libbeat/management/api"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/management"
+)
+
+func init() {
+	management.Register("x-pack", NewConfigManager, feature.Beta)
+}
+
+// ConfigManager handles internal config updates. By retrieving
+// new configs from Kibana and applying them to the Beat
+type ConfigManager struct {
+	config   *Config
+	logger   *logp.Logger
+	client   *api.Client
+	beatUUID uuid.UUID
+	done     chan struct{}
+	registry *reload.Registry
+	wg       sync.WaitGroup
+}
+
+// NewConfigManager returns a X-Pack Beats Central Management manager
+func NewConfigManager(registry *reload.Registry, beatUUID uuid.UUID) (management.ConfigManager, error) {
+	c := defaultConfig()
+	if err := c.Load(); err != nil {
+		return nil, errors.Wrap(err, "reading central management internal settings")
+	}
+	return NewConfigManagerWithConfig(c, registry, beatUUID)
+}
+
+// NewConfigManagerWithConfig returns a X-Pack Beats Central Management manager
+func NewConfigManagerWithConfig(c *Config, registry *reload.Registry, beatUUID uuid.UUID) (management.ConfigManager, error) {
+	var client *api.Client
+	if c.Enabled {
+		var err error
+
+		// Ignore kibana version to avoid permission errors
+		c.Kibana.IgnoreVersion = true
+
+		client, err = api.NewClient(c.Kibana)
+		if err != nil {
+			return nil, errors.Wrap(err, "initializing kibana client")
+		}
+	}
+
+	return &ConfigManager{
+		config:   c,
+		logger:   logp.NewLogger(management.DebugK),
+		client:   client,
+		done:     make(chan struct{}),
+		beatUUID: beatUUID,
+		registry: registry,
+	}, nil
+}
+
+// Enabled returns true if config management is enabled
+func (cm *ConfigManager) Enabled() bool {
+	return cm.config.Enabled
+}
+
+// Start the config manager
+func (cm *ConfigManager) Start() {
+	if !cm.Enabled() {
+		return
+	}
+	cfgwarn.Beta("Central management is enabled")
+	cm.logger.Info("Starting central management service")
+
+	cm.wg.Add(1)
+	go cm.worker()
+}
+
+// Stop the config manager
+func (cm *ConfigManager) Stop() {
+	if !cm.Enabled() {
+		return
+	}
+	cm.logger.Info("Stopping central management service")
+	close(cm.done)
+	cm.wg.Wait()
+}
+
+// CheckRawConfig check settings are correct to start the beat
+func (cm *ConfigManager) CheckRawConfig(cfg *common.Config) error {
```

Comment:
If this function does not return error does it mean the whole configuration is correct? Is this check is going to be superior to the current `test config`?

Reply:
I think I missed a big TODO comment here :smile:

This method is minded to check that the user hasn't configured something that is supposed to be managed my Kibana. In a next PR I will add that code. BTW the `test config` point is good, will need to make sure this check is run there too.

I'm adding a TODO comment now :+1:

Reply:
pushed dbaf5af